File: code/backend/java/core/src/main/java/games/paths/core/persistence/match/LocationEntryStoreAdapter.java
package games.paths.core.persistence.match;

import games.paths.core.entity.match.GamingCharacterInstanceEntity;
import games.paths.core.entity.match.GamingStateLocationsEntityId;
import games.paths.core.entity.match.LogEventsEntity;
import games.paths.core.entity.match.LogMovementEntity;
import games.paths.core.entity.story.LocationEntity;
import games.paths.core.port.match.LocationEntryStorePort;
import games.paths.core.port.story.StoryReadPort;
import games.paths.core.repository.match.GamingCharacterInstanceRepository;
import games.paths.core.repository.match.GamingStateLocationsRepository;
import games.paths.core.repository.match.LogEventsRepository;
import games.paths.core.repository.match.LogMovementRepository;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;

/**
 * LocationEntryStoreAdapter - JPA adapter implementing {@link LocationEntryStorePort}
 * for the Step 33 location engine.
 */
@Repository
@Transactional
public class LocationEntryStoreAdapter implements LocationEntryStorePort {

    private final GamingStateLocationsRepository stateLocationsRepository;
    private final GamingCharacterInstanceRepository characterRepository;
    private final LogEventsRepository logEventsRepository;
    private final LogMovementRepository logMovementRepository;
    private final StoryReadPort storyReadPort;

    public LocationEntryStoreAdapter(GamingStateLocationsRepository stateLocationsRepository,
                                     GamingCharacterInstanceRepository characterRepository,
                                     LogEventsRepository logEventsRepository,
                                     LogMovementRepository logMovementRepository,
                                     StoryReadPort storyReadPort) {
        this.stateLocationsRepository = stateLocationsRepository;
        this.characterRepository = characterRepository;
        this.logEventsRepository = logEventsRepository;
        this.logMovementRepository = logMovementRepository;
        this.storyReadPort = storyReadPort;
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<LocationTriggerView> findLocationTriggers(long idStory, long idLocation) {
        for (LocationEntity l : storyReadPort.findLocationsByStoryId(idStory)) {
            if (l.getId() != null && l.getId().longValue() == idLocation) {
                return Optional.of(new LocationTriggerView(
                        idLocation,
                        l.getIdCard(),
                        l.getIdEventIfFirstTime(),
                        l.getIdEventNotFirstTime(),
                        l.getIdEventIfCharacterEnterEmptyLocation(),
                        l.getIdEventIfCharacterStartTime(),
                        l.getIdEventIfCounterZero(),
                        l.getPriorityAutomaticEvent()));
            }
        }
        return Optional.empty();
    }

    @Override
    @Transactional(readOnly = true)
    public int findFlagVisited(long idMatch, long idLocation) {
        return stateLocationsRepository.findById(new GamingStateLocationsEntityId(idMatch, idLocation))
                .map(s -> nz(s.getFlagVisited()))
                .orElse(0);
    }

    @Override
    public void markStateLocationVisited(long idMatch, long idLocation) {
        stateLocationsRepository.findById(new GamingStateLocationsEntityId(idMatch, idLocation))
                .ifPresent(s -> {
                    if (nz(s.getFlagVisited()) == 1) {
                        return;
                    }
                    s.setFlagVisited(1);
                    stateLocationsRepository.save(s);
                });
    }

    @Override
    @Transactional(readOnly = true)
    public int countOtherCharactersAtLocation(long idMatch, long idLocation, long exceptIdCharacter) {
        int count = 0;
        for (GamingCharacterInstanceEntity c : characterRepository.findByIdMatch(idMatch)) {
            if (c.getId() == null || c.getId() == exceptIdCharacter) {
                continue;
            }
            if (c.getIdLocation() != null && c.getIdLocation().longValue() == idLocation) {
                count++;
            }
        }
        return count;
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<Long> findNominalActorAtLocation(long idMatch, long idLocation) {
        Long lowest = null;
        for (GamingCharacterInstanceEntity c : characterRepository.findByIdMatch(idMatch)) {
            if (c.getId() == null || c.getIdLocation() == null) {
                continue;
            }
            if (c.getIdLocation().longValue() != idLocation) {
                continue;
            }
            if (lowest == null || c.getId() < lowest) {
                lowest = c.getId();
            }
        }
        return Optional.ofNullable(lowest);
    }

    @Override
    public void logAutomaticEvent(long idMatch, Long idCharacter, long idLocation, Long idEvent,
                                  Integer clock, String message) {
        LogEventsEntity e = new LogEventsEntity();
        e.setId(logEventsRepository.findMaxId() + 1);
        e.setIdMatch(idMatch);
        e.setIdCharacterMatch(idCharacter);
        e.setIdLocation(idLocation);
        e.setIdEvent(idEvent);
        e.setClock(clock);
        e.setLogMessage(message);
        logEventsRepository.save(e);
    }

    @Override
    @Transactional(readOnly = true)
    public List<Long> findVisitedLocationIds(long idMatch) {
        Set<Long> ids = new LinkedHashSet<>();
        for (GamingCharacterInstanceEntity c : characterRepository.findByIdMatch(idMatch)) {
            if (c.getIdLocation() != null) {
                ids.add(c.getIdLocation());
            }
        }
        for (LogMovementEntity l : logMovementRepository.findByIdMatch(idMatch)) {
            if (l.getIdLocationFrom() != null) {
                ids.add(l.getIdLocationFrom());
            }
            if (l.getIdLocationTo() != null) {
                ids.add(l.getIdLocationTo());
            }
        }
        return new ArrayList<>(ids);
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<Long> findCharacterLocation(long idMatch, long idCharacter) {
        return characterRepository.findByIdMatchAndId(idMatch, idCharacter)
                .map(GamingCharacterInstanceEntity::getIdLocation);
    }

    private static int nz(Integer v) {
        return v == null ? 0 : v;
    }
}

File: code/backend/python/app/adapters/persistence/match/location_entry_store_adapter.py
"""Step 33 — SQLAlchemy adapter implementing :class:`LocationEntryStorePort`."""
from typing import Any, Dict, List, Optional

from sqlalchemy import func

from app.adapters.persistence.match.models import (
    GamingCharacterInstanceEntity,
    GamingStateLocationEntity,
    LogEventsEntity,
    LogMovementEntity,
)
from app.adapters.persistence.story.models import LocationEntity
from app.adapters.persistence.match.turn_cycle_store_adapter import _new_uuid, _now_iso
from app.core.ports.match.location_entry_ports import LocationEntryStorePort


class LocationEntryStoreAdapter(LocationEntryStorePort):

    def __init__(self, session_factory) -> None:
        self.session_factory = session_factory

    def find_location_triggers(self, id_story: int,
                               id_location: int) -> Optional[Dict[str, Any]]:
        with self.session_factory() as session:
            l = (
                session.query(LocationEntity)
                .filter(LocationEntity.id_story == id_story,
                        LocationEntity.id == id_location)
                .first()
            )
            if l is None:
                return None
            return {
                "id_location": l.id,
                "id_card": l.id_card,
                "id_event_if_first_time": l.id_event_if_first_time,
                "id_event_not_first_time": l.id_event_not_first_time,
                "id_event_if_character_enter_empty_location": l.id_event_if_character_enter_empty_location,
                "id_event_if_character_start_time": l.id_event_if_character_start_time,
                "id_event_if_counter_zero": l.id_event_if_counter_zero,
                "priority_automatic_event": l.priority_automatic_event,
                "key_to_add": l.key_to_add,
                "key_value_to_add": l.key_value_to_add,
                "key_to_add_not_first": l.key_to_add_not_first,
                "key_value_to_add_not_first": l.key_value_to_add_not_first,
            }

    def find_flag_visited(self, id_match: int, id_location: int) -> int:
        with self.session_factory() as session:
            s = self._state_row(session, id_match, id_location)
            return (s.flag_visited or 0) if s is not None else 0

    def mark_state_location_visited(self, id_match: int, id_location: int) -> None:
        with self.session_factory() as session:
            s = self._state_row(session, id_match, id_location)
            if s is None or (s.flag_visited or 0) == 1:
                return
            s.flag_visited = 1
            s.ts_update = _now_iso()
            session.commit()

    def count_other_characters_at_location(self, id_match: int, id_location: int,
                                           except_id_character: int) -> int:
        with self.session_factory() as session:
            return (
                session.query(func.count(GamingCharacterInstanceEntity.id))
                .filter(GamingCharacterInstanceEntity.id_match == id_match,
                        GamingCharacterInstanceEntity.id_location == id_location,
                        GamingCharacterInstanceEntity.id != except_id_character)
                .scalar()
            ) or 0

    def find_nominal_actor_at_location(self, id_match: int,
                                       id_location: int) -> Optional[int]:
        with self.session_factory() as session:
            return (
                session.query(func.min(GamingCharacterInstanceEntity.id))
                .filter(GamingCharacterInstanceEntity.id_match == id_match,
                        GamingCharacterInstanceEntity.id_location == id_location)
                .scalar()
            )

    def log_automatic_event(self, id_match: int, id_character: Optional[int],
                            id_location: int, id_event: Optional[int],
                            clock: Optional[int], message: str) -> None:
        with self.session_factory() as session:
            max_id = session.query(func.max(LogEventsEntity.id)).scalar() or 0
            now = _now_iso()
            session.add(LogEventsEntity(
                id=max_id + 1,
                id_match=id_match,
                uuid=_new_uuid(),
                id_character_match=id_character,
                timestamp=now,
                id_event=id_event,
                id_location=id_location,
                clock=clock,
                log_message=message,
                ts_insert=now,
                ts_update=now,
            ))
            session.commit()

    def find_visited_location_ids(self, id_match: int) -> List[int]:
        with self.session_factory() as session:
            ids: List[int] = []
            for c in (session.query(GamingCharacterInstanceEntity)
                      .filter(GamingCharacterInstanceEntity.id_match == id_match).all()):
                if c.id_location is not None and c.id_location not in ids:
                    ids.append(c.id_location)
            for m in (session.query(LogMovementEntity)
                      .filter(LogMovementEntity.id_match == id_match).all()):
                for loc in (m.id_location_from, m.id_location_to):
                    if loc is not None and loc not in ids:
                        ids.append(loc)
            return ids

    def find_character_location(self, id_match: int,
                                id_character: int) -> Optional[int]:
        with self.session_factory() as session:
            c = (
                session.query(GamingCharacterInstanceEntity)
                .filter(GamingCharacterInstanceEntity.id_match == id_match,
                        GamingCharacterInstanceEntity.id == id_character)
                .first()
            )
            return c.id_location if c is not None else None

    @staticmethod
    def _state_row(session, id_match: int, id_location: int):
        return (
            session.query(GamingStateLocationEntity)
            .filter(GamingStateLocationEntity.id_match == id_match,
                    GamingStateLocationEntity.id_location == id_location)
            .first()
        )
